### PR TITLE
Fix FP S131 (`switch-without-default`): Ignore exhaustive switches on unions and enums

### LINKED ISSUE #3821: Fix FP S131 (`switch-without-default`): Ignore exhaustive switches on unions and enums
https://typescript-eslint.io/rules/switch-exhaustiveness-check

 - No options, no quickfix
 - Replace our implementation if possible

### PR DESCRIPTION
**Pull request by yassin-kammoun-sonarsource**
Fixes #3821

### CODE REVIEW

File: tests/linting/eslint/rules/switch-without-default.test.ts [hunk 1]
```diff
@@ -47,14 +51,121 @@ const tests = {
...
+        switch (x) {
+        }`,
+      errors: [
+        {
+          line: 2,
+          suggestions: [
+            {
+              output: `
+        switch (x) {
+        default: { throw new Error('Not implemented yet'); }
```
Comment (ilia-kebets-sonarsource) [inline tests/linting/eslint/rules/switch-without-default.test.ts:79]: nice fix

File: tests/linting/eslint/rules/switch-without-default.test.ts [hunk 2]
```diff
@@ -47,14 +51,121 @@ const tests = {
...
+        }`,
+            },
+          ],
+        },
+      ],
+    },
+    {
+      code: `
+        type  T = 'foo' | 'bar'; // False negative due to missing type information
```
[the remark this answers is not in the document]
Comment (ilia-kebets-sonarsource) [reply tests/linting/eslint/rules/switch-without-default.test.ts:88]: is it not possible to write a valid test case for your exception?

[the remark this answers is not in the document]
Comment (ilia-kebets-sonarsource) [reply tests/linting/eslint/rules/switch-without-default.test.ts:88]: nevermind

**Comment by yassin-kammoun-sonarsource** [comment]
For the record, I missed it, but I realized that TypeScript ESLint's implementation of the rule already covers the case for enums as well.

File: src/linting/eslint/rules/switch-without-default.ts [hunk 3]
```diff
@@ -19,30 +19,128 @@
...
+ * rule is executed to prevent runtime errors. Furthermore, we need to decorate the rule so that
+ * it raises issues with the same message at the same location, that is, the `switch` keyword.
+ */
+const switchExhaustivenessRule = typeScriptESLintRules['switch-exhaustiveness-check'];
+const decoratedSwitchExhaustivenessRule: Rule.RuleModule = interceptReport(
+  sanitizeTypeScriptESLintRule(switchExhaustivenessRule),
+  function (context: Rule.RuleContext, descriptor: Rule.ReportDescriptor) {
+    const switchNode = (descriptor as any).node.parent as estree.Node;
+    const switchKeyword = getSwitchKeyword(switchNode, context);
+    context.report({ ...descriptor, loc: switchKeyword.loc });
```
[the remark this answers is not in the document]
Comment (yassin-kammoun-sonarsource) [reply src/linting/eslint/rules/switch-without-default.ts:85]: Yes for consistency.

File: src/linting/eslint/rules/switch-without-default.ts [hunk 4]
```diff
@@ -19,30 +19,128 @@
...
+import { sanitizeTypeScriptESLintRule } from '../linter/decoration';
+import { interceptReport, mergeRules } from '../rules/decorators/helpers';
+import { isRequiredParserServices, isUnion } from './helpers';
 
-export const rule: Rule.RuleModule = {
+/**
+ * This rule raises issues on switch statements without a default branch if, and only if,
+ * the discriminant of the switch statement is not of union type. This exception is due to
+ * `switch-exhaustiveness-check` decorated below which checks the exhaustiveness of switch
+ * statements on TypeScript unions. Therefore, we avoid here raising multiple issues if the
```
Comment (ilia-kebets-sonarsource) [inline src/linting/eslint/rules/switch-without-default.ts:33]:
```suggestion
 * statements on TypeScript unions and enums. Therefore, we avoid here raising multiple issues if the
```

File: src/linting/eslint/rules/switch-without-default.ts [hunk 5]
```diff
@@ -19,30 +19,128 @@
...
+ * discriminant of the switch statement denotes a union, provided that type information is available.
```
Comment (ilia-kebets-sonarsource) [inline src/linting/eslint/rules/switch-without-default.ts:34]:
```suggestion
 * discriminant of the switch statement denotes a union or enum, provided that type information is available.
```

**Review by ilia-kebets-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**